RFC-0001 Standard Web5 API Design (APID) Document v0.1.0 (#229)
TBD54566975/web5-rs: docs/API_DESIGN.md

/// Note that this is a subset of the spec.
CLASS Jwk
  /// Identifies the algorithm intended for use with the key.
  PUBLIC DATA alg: string

  /// Represents the key type. e.g. EC for elliptic curve, OKP for Edwards curve
  PUBLIC DATA kty: string

  /// curve name for Elliptic Curve (EC) and Edwards Curve (Ed) keys.
  /// e.g. secp256k1, Ed25519
  PUBLIC DATA crv: string

  /// X coordinate for EC keys, or the public key for OKP.
  PUBLIC DATA x: string

  /// Y coordinate for EC keys.
  PUBLIC DATA y: string?

  /// Private key component for EC or OKP keys.
  PUBLIC DATA d: string?
```

## `InMemoryKeyManager`

```pseudocode!
/// An encapsulation of key material stored in-memory.
CLASS InMemoryKeyManager
  /// Generates new key material and returns the public key represented as a Jwk.
  METHOD generate_key_material(): Jwk

  /// Returns the Ed25519Signer for the given public key.
  METHOD get_signer(public_key: Jwk): Ed25519Signer

  /// For importing keys which may be stored somewhere such as environment variables. Return Jwk is the public key for the given private key.
  METHOD import_key(private_key: Jwk): Jwk
```

> # Digital Signature Algorithm's (DSA)

Reviewer: I'd suggest a more generic name like `Cryptography` or even just `Crypto` since the code here is already (and will continue to be) broader than _signature algorithms_. Key generation now - and in the future things like encryption or key agreement will make this naming obsolete.

Author: The _Custom DSL_ doesn't currently offer prescriptions with respect to namespacing, but I think it may be prudent to consider. I've added namespacing as a missing concept, and I've moved the above section as a sub-section under `Cryptography`, and I've done the same for the VC Data Model 1.1 in your comment below. Thanks!

## `Dsa`

```pseudocode!

> /// The set of Digital Signature Algorithm's natively supported within this SDK.

Reviewer: ```suggestion
/// The set of cryptographic primitives natively supported within this SDK.
```

Reviewer: related to [this comment](https://github.com/TBD54566975/web5-rs/pull/229#discussion_r1638155385) -- this API should follow that logic. it's sensible to have an API for DSA, though it should just be for DSA. Key generation should be separate. Same with encryption in the future - and other cryptographic functions.

Reviewer: which I've noticed you've already done 😄

ENUM Dsa
  Ed25519
```

> We must add support for `Xd25519`, `secp256k1`, and `secp256r1` for [full did:dht conformance](https://did-dht.com/registry/index.html#key-type-index).

## `Signer`

```pseudocode!
/// Set of functionality required to implement to be a compatible DSA signer.
INTERFACE Signer
  /// Signs the given payload by using the encapsulated private key material.
  METHOD sign(payload: []byte): []byte
```

## `Verifier`

```pseudocode!
/// Set of functionality required to implement to be a compatible DSA verifier.
INTERFACE Verifier
  /// Execute the verification of the signature against the message by using the encapsulated public key material.
  METHOD verify(message: []byte, signature: []byte): bool
```

## `Ed25519Generator`

```pseudocode!
/// Generates private key material for Ed25519.
CLASS Ed25519Generator
  /// Generate the private key material; return Jwk includes private key material.

>   STATIC METHOD generate(): Jwk

Author: ```suggestion
  STATIC METHOD generate_key_material(): Jwk
```

?

did_dht.document.alsoKnownAs = "did:example:efgh"
/// Note: you could also add a verification method, set the controller etc.

signer = key_manager.get_signer(public_key)
did_dht.publish(signer)
```

#### Example: Instantiate an existing `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
did_dht = new DidDht(uri)
```

#### Example: Update a `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
did_dht = new DidDht(uri)

/// Set the alsoKnownAs
did_dht.document.alsoKnownAs = "did:example:efgh"
/// Note: you could also add a verification method, set the controller etc.

key_manager = new InMemoryKeyManager()
public_key = key_manager.import_key(private_key) /// assume private_key pre-exists, eg. read from env var
signer = key_manager.get_signer(public_key)

did_dht.publish(signer)
```

#### Example: Resolve a `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
resolution_result = DidDht.resolve(uri)
```

# Verifiable Credential's (VCs)

## `NamedIssuer`

```pseudocode!
CLASS NamedIssuer
  PUBLIC DATA id: string
  PUBLIC DATA name: string
```

> ## `VerifiableCredential`

Reviewer: it may make sense to namespace this, since we'll soon need to account for differnet types of 'verifiable credentials' or the more generic 'digital credentials' - for example: SD-JWT, SD-JWT-VC, mDL/mDoc, W3C VCDM v1 (what we currently support), W3C VCDM v2

Author: Moved under `Data Model 1.1` markdown section

# Web5 API Design (APID) <!-- omit in toc -->

**Last Updated** May 30, 2024

**Version** 0.1.0

**[Custom DSL](./CUSTOM_DSL.md) Version**: 0.1.0

- [Key Material](#key-material)
  - [`Jwk`](#jwk)
  - [`InMemoryKeyManager`](#inmemorykeymanager)
- [Cryptography](#cryptography)

>   - [Digital Signature Algorithm's (DSA)](#digital-signature-algorithms-dsa)

Reviewer 2: ```suggestion
  - [Digital Signature Algorithms (DSA)](#digital-signature-algorithms-dsa)
```

/// Note that this is a subset of the spec.
CLASS Jwk
  /// Identifies the algorithm intended for use with the key.
  PUBLIC DATA alg: string

  /// Represents the key type. e.g. EC for elliptic curve, OKP for Edwards curve
  PUBLIC DATA kty: string

  /// curve name for Elliptic Curve (EC) and Edwards Curve (Ed) keys.
  /// e.g. secp256k1, Ed25519
  PUBLIC DATA crv: string

  /// X coordinate for EC keys, or the public key for OKP.
  PUBLIC DATA x: string

  /// Y coordinate for EC keys.
  PUBLIC DATA y: string?

  /// Private key component for EC or OKP keys.
  PUBLIC DATA d: string?
```

## `InMemoryKeyManager`

```pseudocode!
/// An encapsulation of key material stored in-memory.
CLASS InMemoryKeyManager
  /// Generates new key material and returns the public key represented as a Jwk.
  METHOD generate_key_material(): Jwk

  /// Returns the Ed25519Signer for the given public key.
  METHOD get_signer(public_key: Jwk): Ed25519Signer

  /// For importing keys which may be stored somewhere such as environment variables. Return Jwk is the public key for the given private key.
  METHOD import_key(private_key: Jwk): Jwk
```

# Cryptography

> ## Digital Signature Algorithms (DSA)

Reviewer 3: i see a lot of same concepts here as current impl, but can you tell me how the hierarchy changes with this design?

for example this is how i understand the current impl

![image](https://github.com/TBD54566975/web5-rs/assets/118510103/e1106f1b-07f2-4099-918d-af64047e4542)

Author: The main design choice is to enable features which require cryptographic signing & verification to support both usage of a polymorphic base class called `KeyManager` (& actually in practice, it's a `BearerDid`) but also support direct usage of a DSA implementation by an implementation of the polymorphic base classes for `dsa::Signer` and `dsa::Verifier`. Point being, I can envision plenty of use cases, particularly large-scale enterprise environments, wherein the concept of a `KeyManager` would be considered excessively abstract and instead they need to bring-their-own cryptography through the usage of the DSA polymorphism classes.

This RFC is, of course, slightly out dated at this point, but that was a big part of the original intention.

> ## Digital Signature Algorithm's (DSA)

Reviewer 2: ```suggestion
## Digital Signature Algorithms (DSA)
```

### `Dsa`

```pseudocode!

> /// The set of Digital Signature Algorithm's natively supported within this SDK.

Reviewer 2: ```suggestion
/// The set of Digital Signature Algorithms natively supported within this SDK.
```

ENUM Dsa
  Ed25519
```

> > We must add support for `Xd25519`, `secp256k1`, and `secp256r1` for [full did:dht conformance](https://did-dht.com/registry/index.html#key-type-index).

Author: Oops, that's a typo heh -- it's `X25519`

https://www.rfc-editor.org/rfc/rfc7748#section-5

  /// Spec: https://www.w3.org/TR/did-core/#method-schemes.
  PUBLIC DATA method: string

  /// ID is the method-specific identifier in the DID URI.
  /// Spec: https://www.w3.org/TR/did-core/#method-specific-id.
  PUBLIC DATA id: string

  /// Params is a map containing optional parameters present in the DID URI. These parameters are method-specific.
  /// Spec: https://www.w3.org/TR/did-core/#did-parameters.
  PUBLIC DATA params: Map<string, string>?

  /// Path is an optional path component in the DID URI.
  /// Spec: https://www.w3.org/TR/did-core/#path.
  PUBLIC DATA path: string?

  /// Query is an optional query component in the DID URI, used to express a request for a specific representation or resource related to the DID.
  /// Spec: https://www.w3.org/TR/did-core/#query.
  PUBLIC DATA query: string?

  /// Fragment is an optional fragment component in the DID URI, used to reference a specific part of a DID document.
  /// Spec: https://www.w3.org/TR/did-core/#fragment.
  PUBLIC DATA fragment: string?

  CONSTRUCTOR(uri: string)
```

### Example: Instantiate from a `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"

> did = new Did(uri)

Reviewer 3: what can i do with this `Did`? usually i create a bearerdid so i can sign over something (i.e. a message)

Reviewer 3: it just turns the did string to an object?

Author: this parses the DID URI into an object representing the Identifier from [the DID Core spec](https://www.w3.org/TR/did-core/#identifier), so it's use cases would be, for example, dereferencing wherein a framgment is needed to be parsed out. It's functionally equivalent to the concept of a `Url.parse(http_endpoint)` which many languages natively support, but specific to DID URI's

/// The error code from the resolution process.
ENUM ResolutionMetadataError
  invalidDid
  notFound
  representationNotSupported
  methodNotSupported
  invalidDidDocument
  invalidDidDocumentLength
  internalError
```

### `ResolutionMetadata`

```pseudocode!
/// Metadata about the given resolution.
CLASS ResolutionMetadata
  /// The error code from the resolution process.
  PUBLIC DATA error: ResolutionMetadataError?
```

### `DocumentMetadata`

```pseudocode!
/// Metadata about the DID Document.
CLASS DocumentMetadata
  /// 🚧
  PUBLIC DATA created: string?

  /// 🚧
  PUBLIC DATA updated: string?

  /// 🚧
  PUBLIC DATA deactivated: bool?

  /// 🚧
  PUBLIC DATA nextUpdate: string?

  /// 🚧
  PUBLIC DATA versionId: string?

  /// 🚧
  PUBLIC DATA nextVersionId: string?

  /// 🚧
  PUBLIC DATA equivalentId: []string?

  /// 🚧
  PUBLIC DATA canonicalId: string?
```

## Methods

### `DidJwk`

```pseudocode!
CLASS DidJwk
  PUBLIC DATA did: Did
  PUBLIC DATA document: Document
  CONSTRUCTOR(public_key: Jwk)
  CONSTRUCTOR(uri: string)
  STATIC METHOD resolve(uri: string): ResolutionResult
```

#### Example: Create a `did:jwk`

```pseudocode!
key_manager = new InMemoryKeyManager()

> public_key = key_manager.generate_key_material()

Reviewer 3: can you tell me a bit more about what this function does? would an equivalent in the current impl be `KeyManager#GeneratePrivateKey()` which returns the keyid after storing the private key in the keystore (but now you're returning the public key as the keyid equivalent) ?

Author: Yeah it's functionally equivalent

This was subsequently removed in favor of:

```pseudocode!
let key_manager = new InMemoryKeyManager();
let public_jwk = key_manager.import_private_jwk(Ed25519Generator::generate())
```

Author: And yeah it's potay-to potah-to with regards to a string representing the thumbprint of the public key; both are functionally equivalent. I chose to go with the JWK being the point of reference because it's one less abstraction to accidentally break.

did_jwk = new DidJwk(public_key)
```

#### Example: Instantiate an existing `did:jwk`

```pseudocode!
uri = "did:jwk:eyJrdHkiOiJFQyIsInVzZSI6InNpZyIsImNydiI6InNlY3AyNTZrMSIsImtpZCI6ImkzU1BSQnRKS292SEZzQmFxTTkydGk2eFFDSkxYM0U3WUNld2lIVjJDU2ciLCJ4IjoidmRyYnoyRU96dmJMRFZfLWtMNGVKdDdWS04VEZaTm1BOVlnV3p2aGg3VSIsInkiOiJWTEZxUU1aUF9Bc3B1Y1hvV1gyLWJHWHBBTzFmUTVMbjE5VjVSQXhyZZVIiwiYWxnIjoiRVMyNTZLIn0"

> did_jwk = new DidJwk(uri)

Reviewer 3: this is only possible because it's a DidJwk, right?

```

#### Example: Resolve a `did:jwk`

```pseudocode!
uri = "did:jwk:eyJrdHkiOiJFQyIsInVzZSI6InNpZyIsImNydiI6InNlY3AyNTZrMSIsImtpZCI6ImkzU1BSQnRKS292SEZzQmFxTTkydGk2eFFDSkxYM0U3WUNld2lIVjJDU2ciLCJ4IjoidmRyYnoyRU96dmJMRFZfLWtMNGVKdDdWS04VEZaTm1BOVlnV3p2aGg3VSIsInkiOiJWTEZxUU1aUF9Bc3B1Y1hvV1gyLWJHWHBBTzFmUTVMbjE5VjVSQXhyZZVIiwiYWxnIjoiRVMyNTZLIn0"
resolution_result = DidJwk.resolve(uri)
```

### `DidWeb`

```pseudocode!
CLASS DidWeb
  CONSTRUCTOR(uri: string)
  STATIC METHOD resolve(uri: string): ResolutionResult
```

#### Example: Instantiate an existing `did:web`

```pseudocode!
uri = "did:web:example.com"
did_web = new DidWeb(uri)
```

#### Example: Resolve a `did:web`

```pseudocode!
uri = "did:web:example.com"
resolution_result = DidWeb.resolve(uri)
```

### `DidDht`

```pseudocode!
CLASS DidDht
  PUBLIC DATA did: Did
  PUBLIC DATA document: Document
  CONSTRUCTOR(identity_key: Jwk)

>   CONSTRUCTOR(uri: string)

Reviewer 3: how could i create a DidDht with knowing just the uri?

Reviewer 3: central to what i want to ask here is 

> what can i _do_ with a DidWhateverMethod once instantiated?

Author: The `DidDht` is an encapsulation around the `Did` and `Document` abstraction, both of those are accessible via the DID URI. Admittedly I'm not sure this has a concrete use case right now 🤔

Author: > in my mind, DIDs (perhaps it's more BearerDids) serve one purpose - to be a vessel that identifies me and holds a signing key so i can prove to bob that a thing i sent him actually came from me

Yeah you're referring specifically to bearer DID's, which is our own concept. The concept of a DID is actually moreso on the matter of accessing public keys, so for example if one wanted to verify a cryptographic signature and they had a DID URI, they could call this to gain access to the public key for verification.

  METHOD publish(signer: Signer)
  METHOD deactivate(signer: Signer)
  STATIC METHOD resolve(uri: string): ResolutionResult
```

> [!NOTE]
> `resolve()` makes use of [`Ed25519Verifier`](#ed25519verifier) internally for DNS packet verification.

#### Example: Create & publish a `did:dht`

```pseudocode!
key_manager = new InMemoryKeyManager()
public_key = key_manager.generate_key_material()
did_dht = new DidDht(public_key)
signer = key_manager.get_signer(public_key)
did_dht.publish(signer)
```

#### Example: Create a `did:dht`, add to the Core Properties & publish

> [!NOTE]
> The call to the `new DidDht()` constructor only adds the minimum requirements to the DID Document.
> If additional [Core Properties](https://www.w3.org/TR/did-core/#core-properties) are required, update the `document` data member prior-to the call to `publish()`.

```pseudocode!
key_manager = new InMemoryKeyManager()
public_key = key_manager.generate_key_material()
did_dht = new DidDht(public_key)

/// Set the alsoKnownAs

> did_dht.document.alsoKnownAs = "did:example:efgh"

Reviewer 3: oof. this is not easy to do in all languages (we've already run into this issue in kotlin)

Author: You're referring to instantiating an instance and subsequently mutating it? What do you mean by "easy?"

/// Note: you could also add a verification method, set the controller etc.

signer = key_manager.get_signer(public_key)
did_dht.publish(signer)
```

#### Example: Instantiate an existing `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
did_dht = new DidDht(uri)
```

#### Example: Update a `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
did_dht = new DidDht(uri)

/// Set the alsoKnownAs
did_dht.document.alsoKnownAs = "did:example:efgh"
/// Note: you could also add a verification method, set the controller etc.

key_manager = new InMemoryKeyManager()
public_key = key_manager.import_key(private_key) /// assume private_key pre-exists, eg. read from env var
signer = key_manager.get_signer(public_key)

did_dht.publish(signer)
```

#### Example: Resolve a `did:dht`

```pseudocode!
uri = "did:dht:i9xkp8ddcbcg8jwq54ox699wuzxyifsqx4jru45zodqu453ksz6y"
resolution_result = DidDht.resolve(uri)
```

> # Verifiable Credential's (VCs)

Reviewer: ```suggestion
# Verifiable Credentials (VCs)
```

## Data Model 1.1

### `NamedIssuer`

```pseudocode!
CLASS NamedIssuer
  PUBLIC DATA id: string
  PUBLIC DATA name: string
```

### `VerifiableCredential`

> [!WARNING]
> The following is incomplete in that an `Object` is not currently supported in the Custom DSL; the matter of the `Object` below is a placeholder and expected to be completed in a subsequent version.

> [!WARNING]
> We need to consider default behaviors such as always including the base `@context` and `type`

```pseudocode!
CLASS VerifiableCredential
  PUBLIC DATA @context: []string
  PUBLIC DATA id: string
  PUBLIC DATA type: []string
  PUBLIC DATA issuer: string | NamedIssuer
  PUBLIC DATA issuanceDate: string
  PUBLIC DATA expirationDate: string?
  PUBLIC DATA credentialSubject: Object  # 🚧 `Object` not supported 🚧
  CONSTRUCTOR(context: []string, id: string, type: []string, issuer: string | NamedIssuer, issuanceDate: string, expirationDate: string?)
  CONSTRUCTOR(vcjwt: string)
  CONSTRUCTOR(vcjwt: string, verifier: Verifier)

>   METHOD sign(signer: Signer): string

Reviewer 3: are you missing a `STATIC METHOD verify(vcjwt: string)` here?

Author: `CONSTRUCTOR(vcjwt: string)` executes cryptographic verification. We're in kind of a halfsy-state wherein I have some `STATIC METHOD`'s but I'm slightly skeptical to full-commit to the concept of a static method because the concept varies from language to language and IMO we misuse the concept in a number of our implementations.
